### PR TITLE
[inductor] Add logging for distributed collective ops for multi‑rank diagnostics

### PR DESCRIPTION
**Pull request by skarjala**
This change introduces structured logging of the collective communication schedule, enabling downstream tools (e.g. TLParse) to ingest and analyze per‑rank collective‐order information for multi‑rank jobs.

- Iterates over scheduler.nodes, filters for _CollectiveKernel nodes
- Extracts each op’s python_kernel_name
- Emits a structured JSON payload under the inductor_collective_schedule artifact name
- Dumps the full schedule list to collective_schedule.json via the PyTorch trace‑structured artifact
- Added comprehensive unit tests for collective schedule tracing: Created test_collective_schedule_empty() and test_collective_schedule_real() tests to verify structured trace logging works correctly for both empty collective schedules and real collective operations (like all_reduce and wait_tensor from _c10d_functional ops).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #159190

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

### CODE REVIEW

File: torch/_inductor/debug.py [hunk 1]
```diff
@@ -46,6 +48,7 @@
 
 ir_pre_fusion_log = getArtifactLogger(__name__, "ir_pre_fusion")
 ir_post_fusion_log = getArtifactLogger(__name__, "ir_post_fusion")
+collective_schedule_log = getArtifactLogger(__name__, "collective_schedule")
```
Comment (yushangdi) [inline torch/_inductor/debug.py]: where is this used?

File: torch/_inductor/debug.py [hunk 2]
```diff
@@ -678,6 +681,21 @@ def build_node_info(node: ir.IRNode) -> dict[str, str]:
                 json.dump(info_dict, fd)
                 fd.write("\n")
 
+    def dump_collective_schedule(self, schedule: list[dict[str, Any]]) -> None:
+        with self.fopen("collective_schedule.json", "w", encoding="utf-8") as fd:
```
Comment (yushangdi) [inline torch/_inductor/debug.py]: You don't need to dump this json here if you just want the result in tlparse. Why do we need to dump it to a json here?

File: torch/_inductor/config.py [hunk 3]
```diff
@@ -1792,6 +1792,9 @@ class trace:
         or os.environ.get("INDUCTOR_PROVENANCE", "0") == "1"
     )
 
+    log = os.environ.get("INDUCTOR_TRACE_LOG", "1" if enabled else "0") == "1"
+    dump_collective_schedule = True
```
Comment (yushangdi) [inline torch/_inductor/config.py]: If you're going to dump this by default, I don't think you need this flag. You can just move `dump_collective_schedule` out of the DebugContext, and put it in debug.py without putting it in any class.

File: torch/_inductor/config.py [hunk 4]
```diff
@@ -1792,6 +1792,9 @@ class trace:
         or os.environ.get("INDUCTOR_PROVENANCE", "0") == "1"
     )
 
+    log = os.environ.get("INDUCTOR_TRACE_LOG", "1" if enabled else "0") == "1"
```
Comment (yushangdi) [inline torch/_inductor/config.py]: do we use this flag?

File: torch/_inductor/debug.py (hunk 2 above)
Comment (yushangdi) [reply torch/_inductor/debug.py]: one reason to dump the result as a json is to write test case for it. It's a little tricky to write test case for the dumped  traced_structure results.

You can follow the examples in https://github.com/pytorch/pytorch/blob/main/test/inductor/test_provenance_tracing.py and write some test cases for the dumped information.

Comment (skarjala) [reply torch/_inductor/debug.py]: If it’s not needed I think I’ll just remove it, I was only using it for debugging purposes.

Comment (yushangdi) [reply torch/_inductor/debug.py]: > If it’s not needed I think I’ll just remove it, I was only using it for debugging purposes.

sure, but let's add some unit tests

File: torch/_inductor/debug.py (hunk 1 above)
Comment (skarjala) [reply torch/_inductor/debug.py]: We dont need it, fixed.

File: torch/_inductor/debug.py (hunk 2 above)
Comment (skarjala) [reply torch/_inductor/debug.py]: So should I re-add the json dump here so I can add unit tests?

File: torch/_inductor/config.py (hunk 4 above)
Comment (skarjala) [reply torch/_inductor/config.py]: Fixed, now we just dump by default.

File: torch/_inductor/debug.py [hunk 5]
```diff
@@ -37,6 +38,7 @@
     FusedSchedulerNode,
     NopKernelSchedulerNode,
     OutputNode,
+    Scheduler,
```
Comment (xmfan) [inline torch/_inductor/debug.py]: if the collective schedule information is always contained within the nodes, maybe just pass in the nodes?

File: torch/_inductor/debug.py [hunk 6]
```diff
@@ -693,6 +695,30 @@ def log_ir_post_fusion(nodes: SchedulerNodeList) -> None:
     V.debug.ir_post_fusion(nodes)
 
 
+def dump_collective_schedule(schedule: list[dict[str, Any]]) -> None:
+    try:
+        trace_structured_artifact(
+            "inductor_collective_schedule",
+            "string",  # encoding
```
Comment (xmfan) [inline torch/_inductor/debug.py]: Here's an example for json encoding, which will make your parsing easier as we add more information to this log file: 
https://github.com/pytorch/pytorch/blob/3db8623dcba7c46ceed5bafc8f43cdd99c79c89d/torch/_dynamo/compiled_autograd.py#L422-L429

File: torch/_inductor/debug.py [hunk 7]
```diff
@@ -693,6 +695,30 @@ def log_ir_post_fusion(nodes: SchedulerNodeList) -> None:
...
+
+
+def log_collective_schedule(scheduler: "Scheduler") -> None:
+    schedule = [
+        {"op_name": getattr(op, "python_kernel_name", None)}
+        for node in scheduler.nodes
+        if isinstance(op := getattr(node, "node", None), ir._CollectiveKernel)
+    ]
+
+    dump_collective_schedule(schedule)
```
Comment (xmfan) [inline torch/_inductor/debug.py]: nit: if this function is meant to only be used in the local file, typically we prefix its name with underscore
```suggestion
    _dump_collective_schedule(schedule)
```

File: torch/_logging/_registrations.py [hunk 8]
```diff
@@ -181,6 +181,11 @@
     "Inductor scheduler information. Useful if working on Inductor fusion algo",
     off_by_default=True,
 )
+register_artifact(
```
Comment (xmfan) [inline torch/_logging/_registrations.py]: you won't need this, if you use the API i mentioned above for json encoding

File: torch/_inductor/debug.py (hunk 2 above)
Comment (xmfan) [reply torch/_inductor/debug.py]: you've added unit tests for tlparse artifacts before (for a different problem), take a look again at your starter task PRs

File: torch/_inductor/debug.py (hunk 7 above)
Comment (skarjala) [reply torch/_inductor/debug.py]: Fixed

File: torch/_logging/_registrations.py (hunk 8 above)
Comment (skarjala) [reply torch/_logging/_registrations.py]: Fixed

**Comment by xmfan** [comment]
Take a look at the CI failures. Most of them just look like they need a EXPECTTEST_ACCEPT=1, also you need unit tests for this PR

File: test/dynamo/test_structured_trace.py [hunk 9]
```diff
@@ -1120,6 +1108,80 @@ def user_context() -> str:
...
+
+        from torch._inductor.debug import log_collective_schedule
+
+        log_collective_schedule([])
+
+        self.assertIn('"inductor_collective_schedule"', self.buffer.getvalue())
+        self.assertEqual(json.loads(payload_buffer.getvalue()), [])
+        self.assertParses()
+
+        trace_log.removeHandler(payload_handler)
```
Comment (yushangdi) [inline test/dynamo/test_structured_trace.py]: it's better to write _setup_collective_schedule_capture() as a context manager and wrap the test inside the context, instead of manually removing the handler at the end

File: torch/_inductor/debug.py [hunk 10]
```diff
@@ -693,6 +694,33 @@ def log_ir_post_fusion(nodes: SchedulerNodeList) -> None:
...
+    except Exception:
+        log.debug(
+            "Failed to log inductor_collective_schedule via structured logging",
+            exc_info=True,
+        )
+
+
+def log_collective_schedule(nodes: Sequence[BaseSchedulerNode]) -> None:
+    schedule = [
+        {"op_name": getattr(op, "python_kernel_name", None)}
```
Comment (yushangdi) [inline torch/_inductor/debug.py]: why do we need the "op_name" key here? seems like a list of names would give the same information?

Comment (skarjala) [reply torch/_inductor/debug.py]: Fixed to just use getattr(op, "python_kernel_name", None)

File: test/dynamo/test_structured_trace.py [hunk 11]
```diff
@@ -1156,30 +1184,30 @@ def forward(self, x):
...
+                payload_content = payload_buffer.getvalue().strip()
+                if payload_content:
+                    schedule = json.loads(payload_content)
+                    self.assertIsInstance(schedule, list)
 
-                # Verify expected collective operations are present
-                op_names = " ".join(op.get("op_name", "") for op in schedule)
+                    # Verify expected collective operations are present
+                op_names = " ".join(op or "" for op in schedule)
                 self.assertIn("all_reduce", op_names)
```
Comment (yushangdi) [inline test/dynamo/test_structured_trace.py]: nit: can we just do self.assertIn("all_reduce", schedule)?

otherwise LGTM.

File: test/dynamo/test_structured_trace.py [hunk 12]
```diff
@@ -1120,6 +1138,81 @@ def user_context() -> str:
...
+
+        try:
+            with self._setup_collective_schedule_capture() as payload_buffer:
+                torch._dynamo.reset()
+
+                mod = CollectiveModule().eval()
+                compiled = torch.compile(mod, backend="inductor")
+
+                with torch.no_grad():
+                    compiled(torch.randn(4, 4))
```
Comment (xmfan) [inline test/dynamo/test_structured_trace.py]: .eval() and torch.no_grad() are unnecessary if none of your tensors requires_grad

File: test/dynamo/test_structured_trace.py [hunk 13]
```diff
@@ -1120,6 +1138,81 @@ def user_context() -> str:
...
+
+                # Verify collective schedule artifact was logged
+                self.assertIn('"inductor_collective_schedule"', self.buffer.getvalue())
+
+                payload_content = payload_buffer.getvalue().strip()
+                if payload_content:
+                    schedule = json.loads(payload_content)
+                    self.assertIsInstance(schedule, list)
```
Comment (xmfan) [inline test/dynamo/test_structured_trace.py]: should this test be passing for the else branch?

Comment (skarjala) [reply test/dynamo/test_structured_trace.py]: It shouldnt, removed the conditional check now

File: test/dynamo/test_structured_trace.py (hunk 12 above)
Comment (skarjala) [reply test/dynamo/test_structured_trace.py]: fixed

File: test/dynamo/test_structured_trace.py [hunk 14]
```diff
@@ -1120,6 +1129,79 @@ def user_context() -> str:
...
+
+            log_collective_schedule([])
+
+            self.assertIn('"inductor_collective_schedule"', self.buffer.getvalue())
+            self.assertEqual(json.loads(payload_buffer.getvalue()), [])
+            self.assertParses()
+
+    @requires_tlparse
+    @requires_distributed()
+    @torch._inductor.config.patch("fx_graph_cache", False)
```
Comment (xmfan) [inline test/dynamo/test_structured_trace.py:1169]:
```suggestion
    @torch._inductor.config.patch("force_disable_caches", True)
```

File: test/dynamo/test_structured_trace.py [hunk 15]
```diff
@@ -1120,6 +1129,79 @@ def user_context() -> str:
...
+                schedule = json.loads(payload_content)
+                self.assertIsInstance(schedule, list)
+
+                # Verify expected collective operations are present
+                self.assertIn(
+                    "torch.ops._c10d_functional.all_reduce_.default", schedule
+                )
+                self.assertIn(
+                    "torch.ops._c10d_functional.wait_tensor.default", schedule
+                )
```
Comment (xmfan) [inline test/dynamo/test_structured_trace.py]: you should try out self.assertExpectedInline on str(schedule)!

**Review by yushangdi — APPROVED** [review]

**Review by xmfan — APPROVED** [review]
CI failure looks strange, maybe rebase

**Comment by malfet** [comment]
@skarjala why do you think failure was unrelated to your change?

File: test/dynamo/test_structured_trace.py [hunk 16]
```diff
@@ -21,6 +22,7 @@
 from torch._logging._internal import TorchLogsFormatter
 from torch.nn.parallel import DistributedDataParallel as DDP
 from torch.testing._internal.common_utils import find_free_port
+from torch.testing._internal.distributed.fake_pg import FakeStore
```
Comment (xmfan) [inline test/dynamo/test_structured_trace.py]: not all builds have distributed! look back at how you added it for your FakePG starter task
